I was able to build Discord on Darwin with the Moonlight package by just
repurposing the `postInstall` script for Vencord, but I got an error when
starting Discord:

```console
Uncaught Exception:
Error: Cannot find module '/nix/store/46i8kpi56l31b66lb35hlnznm5ngyr4s-discord-0.0.333/Applications/Discord.app/Contents/Resources/app.asar/injector.js'
at Module._resolveFilename (node:internal/modules/cjs/loader:1232:15)
at s._resolveFilename (node:electron/js2c/browser_init:2:124479)
at node:electron/js2c/browser_init:2:130101
at node:electron/js2c/browser_init:2:130350
at node:electron/js2c/browser_init:2:130354
at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/realm:398:7)
at BuiltinModule.compileForPublicLoader (node:internal/bootstrap/realm:337:10)
at loadBuiltinModule (node:internal/modules/helpers:104:7)
at Module._load (node:internal/modules/cjs/loader:1083:17)
at c._load (node:electron/js2c/node_init:2:16955)
```

Used:
```nix
+ lib.strings.optionalString withMoonlight ''
  mv $out/Applications/${desktopName}.app/Contents/Resources/app.asar $out/Applications/${desktopName}.app/Contents/Resources/_app.asar
  mkdir $out/Applications/${desktopName}.app/Contents/Resources/app.asar
  echo '{"name":"discord","main":"injector.js","private": true}' > $out/Applications/${desktopName}.app/Contents/Resources/app.asar/package.json
  echo 'require("${moonlight}/injector.js").inject(require("path").join(__dirname, "../_app.asar"));' > $out/Applications/${desktopName}.app/Contents/Resources/app.asar/index.js
'';
```

--- a/maintainers/maintainer-list.nix
+++ b/maintainers/maintainer-list.nix
@@ -9885,6 +9885,11 @@
     github = "ilyakooo0";
     githubId = 6209627;
   };
+  ilys = {
+    name = "ilys";
+    github = "ilyist";
+    githubId = 185637611;
+  };
   imadnyc = {
     email = "me@imad.nyc";
     github = "imadnyc";

--- a/pkgs/applications/networking/instant-messengers/discord/darwin.nix
+++ b/pkgs/applications/networking/instant-messengers/discord/darwin.nix
@@ -17,7 +17,13 @@
   openasar,
   withVencord ? false,
   vencord,
+  withMoonlight ? false,
+  moonlight,
 }:
+
+assert lib.assertMsg (
+  !(withMoonlight && withVencord)
+) "discord: Moonlight and Vencord can not be enabled at the same time";
 
 let
   disableBreakingUpdates =
@@ -72,6 +78,13 @@ stdenv.mkDerivation {
       mkdir $out/Applications/${desktopName}.app/Contents/Resources/app.asar
       echo '{"name":"discord","main":"index.js"}' > $out/Applications/${desktopName}.app/Contents/Resources/app.asar/package.json
       echo 'require("${vencord}/patcher.js")' > $out/Applications/${desktopName}.app/Contents/Resources/app.asar/index.js
+    ''
+
+    + lib.strings.optionalString withMoonlight ''
+      mv $out/Applications/${desktopName}.app/Contents/Resources/app.asar $out/Applications/${desktopName}.app/Contents/Resources/_app.asar
+      mkdir $out/Applications/${desktopName}.app/Contents/Resources/app.asar
+      echo '{"name":"discord","main":"injector.js","private": true}' > $out/Applications/${desktopName}.app/Contents/Resources/app.asar/package.json
+      echo 'require("${moonlight}/injector.js").inject(require("path").join(__dirname, "../_app.asar"));' > $out/Applications/${desktopName}.app/Contents/Resources/app.asar/injector.js
     '';
 
   passthru = {

--- a/pkgs/applications/networking/instant-messengers/discord/linux.nix
+++ b/pkgs/applications/networking/instant-messengers/discord/linux.nix
@@ -62,8 +62,14 @@
   openasar,
   withVencord ? false,
   vencord,
+  withMoonlight ? false,
+  moonlight,
   withTTS ? true,
 }:
+
+assert lib.assertMsg (
+  !(withMoonlight && withVencord)
+) "discord: Moonlight and Vencord can not be enabled at the same time";
 
 let
   disableBreakingUpdates =
@@ -195,6 +201,12 @@ stdenv.mkDerivation rec {
       mkdir $out/opt/${binaryName}/resources/app.asar
       echo '{"name":"discord","main":"index.js"}' > $out/opt/${binaryName}/resources/app.asar/package.json
       echo 'require("${vencord}/patcher.js")' > $out/opt/${binaryName}/resources/app.asar/index.js
+    ''
+    + lib.strings.optionalString withMoonlight ''
+      mv $out/opt/${binaryName}/resources/app.asar $out/opt/${binaryName}/resources/_app.asar
+      mkdir $out/opt/${binaryName}/resources/app
+      echo '{"name":"discord","main":"injector.js","private": true}' > $out/opt/${binaryName}/resources/app/package.json
+      echo 'require("${moonlight}/injector.js").inject(require("path").join(__dirname, "../_app.asar"));' > $out/opt/${binaryName}/resources/app/injector.js
     '';
 
   desktopItem = makeDesktopItem {

--- a/pkgs/by-name/mo/moonlight/package.nix
+++ b/pkgs/by-name/mo/moonlight/package.nix
@@ -1,0 +1,66 @@
+{
+  lib,
+  stdenv,
+  nodejs,
+  pnpm_9,
+  fetchFromGitHub,
+  nix-update-script,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "moonlight";
+  version = "1.3.9";
+
+  src = fetchFromGitHub {
+    owner = "moonlight-mod";
+    repo = "moonlight";
+    tag = "v${finalAttrs.version}";
+    hash = "sha256-WhPQ7JYfE8RBhDknBunKdW1VBxrklb3UGnMgk5LFVFA=";
+  };
+
+  nativeBuildInputs = [
+    nodejs
+    pnpm_9.configHook
+  ];
+
+  pnpmDeps = pnpm_9.fetchDeps {
+    inherit (finalAttrs) pname version src;
+
+    hash = "sha256-KZFHcW/OVjTDXZltxPYGuO+NWjuD5o6HE/E9JQZmrG8=";
+  };
+
+  buildPhase = ''
+    runHook preBuild
+
+    pnpm run build
+
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    cp -r dist $out
+
+    runHook postInstall
+  '';
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = with lib; {
+    description = "Discord client modification, focused on enhancing user and developer experience";
+    longDescription = ''
+      Moonlight is a ***passion project***—yet another Discord client mod—focused on providing a decent user
+      and developer experience. Heavily inspired by hh3 (a private client mod) and the projects before it, namely EndPwn.
+      All core code is original or used with permission from their respective authors where not copyleft.
+    '';
+    homepage = "https://moonlight-mod.github.io";
+    changelog = "https://github.com/moonlight-mod/moonlight/blob/main/CHANGELOG.md";
+
+    license = licenses.lgpl3;
+    maintainers = with maintainers; [
+      ilys
+      donteatoreo
+    ];
+  };
+})

